File: src/sonar_data_pipeline/sonar_publisher.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float32

class SonarPublisher(Node):
    def __init__(self):
        super().__init__('sonar_publisher')
        self.publisher_ = self.create_publisher(Float32, 'sonar/range', 10)
        self.timer = self.create_timer(1.0, self.timer_callback)

    def timer_callback(self):
        msg = Float32()
        msg.data = 15.5  # Mock distance to target in meters
        self.publisher_.publish(msg)
        self.get_logger().info(f'Publishing Sonar Range: {msg.data} m')

def main(args=None):
    rclpy.init(args=args)
    node = SonarPublisher()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/sonar_data_pipeline/src/sonar_publisher.cpp
#include <chrono>
#include <memory>
#include <string>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class SonarPublisher : public rclcpp::Node {
public:
    SonarPublisher() : Node("sonar_publisher"), count_(0) {
        publisher_ = this->create_publisher<std_msgs::msg::String>("sonar_data", 10);
        timer_ = this->create_wall_timer(
            1s, std::bind(&SonarPublisher::timer_callback, this));
    }

private:
    void timer_callback() {
        auto message = std_msgs::msg::String();
        message.data = "Sonar Ping Distance: " + std::to_string(count_++) + "m";
        RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
        publisher_->publish(message);
    }
    
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    size_t count_;
};

int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SonarPublisher>());
    rclcpp::shutdown();
    return 0;
}
